File: edulens-agents/tools/memory_tools.py
"""Shared tool: memory retrieval (mock version).
In production, replace with real AgentCore Memory SDK calls."""

import json

from strands import tool

from .mock_data import MOCK_MEMORY_RECORDS


@tool
def retrieve_memories(query: str, namespace: str = "", max_results: int = 5) -> str:
    """Search past conversation memories for relevant context. Returns long-term memory records matching the query.

    Args:
        query: Search query for memory retrieval.
        namespace: Optional namespace filter (e.g. 'parent-conversations', 'tutoring-sessions').
        max_results: Max results to return (default 5).
    """
    query_lower = query.lower()
    words = query_lower.split()

    scored = []
    for record in MOCK_MEMORY_RECORDS:
        if namespace and record["namespace"] != namespace:
            continue
        content_lower = record["content"].lower()
        match_count = sum(1 for w in words if w in content_lower)
        score = match_count / len(words) if words else 0
        if score > 0:
            scored.append({
                "content": record["content"],
                "relevance": f"{score:.2f}",
                "namespace": record["namespace"],
            })

    scored.sort(key=lambda x: float(x["relevance"]), reverse=True)

    return json.dumps({
        "query": query,
        "resultCount": len(scored[:max_results]),
        "records": scored[:max_results],
    }, indent=2)

File: edulens-agents-v2/src/tools/memory-tools.ts
/**
 * Memory Tools - Simplified version for compilation
 */

import { tool } from '@strands-agents/sdk';
import { z } from 'zod';

// Zod schemas for memory operations
const SaveMemorySchema = z.object({
  content: z.string().min(1, 'Memory content is required'),
  namespace: z.enum(['conversations', 'preferences', 'insights', 'progress', 'sessions']),
  actorId: z.string().min(1, 'Actor ID is required for namespace isolation')
});

const RetrieveMemorySchema = z.object({
  query: z.string().min(1, 'Search query is required'),
  actorId: z.string().min(1, 'Actor ID is required for namespace isolation'),
  limit: z.number().int().min(1).max(20).optional().default(5)
});

/**
 * Save memory to AgentCore Memory with proper namespace isolation
 *
 * TODO: Replace mock with AgentCore Memory SDK when available:
 *   import { MemoryClient } from 'bedrock-agentcore/memory';
 *   const memoryClient = new MemoryClient({ region: 'us-west-2' });
 *   await memoryClient.putMemory({
 *     namespace: namespacePath,
 *     content,
 *     metadata: { actorId, timestamp: new Date().toISOString() }
 *   });
 */
async function saveMemory(input: z.infer<typeof SaveMemorySchema>): Promise<string> {
  const { content, namespace, actorId } = input;

  // Create namespace path with actor isolation
  const namespacePath = `/actors/${actorId}/${namespace}`;

  // TODO: Replace with memoryClient.putMemory({ namespace: namespacePath, content })
  const result = {
    success: true,
    memoryId: `mem_${Date.now()}_${Math.random().toString(36).substring(7)}`,
    namespace: namespacePath,
    actorId,
    contentLength: content.length
  };

  return JSON.stringify(result, null, 2);
}

/**
 * Retrieve relevant memories using semantic search
 *
 * TODO: Replace mock with AgentCore Memory SDK when available:
 *   const results = await memoryClient.searchMemory({
 *     namespace: `/actors/${actorId}`,
 *     query,
 *     maxResults: limit,
 *     searchType: 'semantic'
 *   });
 *   return JSON.stringify({ query, actorId, resultsCount: results.length, memories: results });
 */
async function retrieveMemory(input: z.infer<typeof RetrieveMemorySchema>): Promise<string> {
  const { query, actorId, limit } = input;

  // TODO: Replace with memoryClient.searchMemory({ namespace, query, maxResults: limit })
  const mockMemories = [
    {
      id: 'mem_001',
      content: 'Student showed strong pattern recognition in math during previous session.',
      namespace: `/actors/${actorId}/conversations`,
      similarity: 0.85
    },
    {
      id: 'mem_002',
      content: 'Parent expressed concern about test anxiety during last conversation.',
      namespace: `/actors/${actorId}/insights`,
      similarity: 0.72
    }
  ];

  const searchResults = mockMemories.slice(0, limit);

  const result = {
    query,
    actorId,
    resultsCount: searchResults.length,
    memories: searchResults
  };

  return JSON.stringify(result, null, 2);
}

// Tool definitions using Strands SDK
export const memoryTools = [
  tool({
    name: 'save_memory',
    description: 'Save important information to long-term memory with proper actor namespace isolation.',
    inputSchema: SaveMemorySchema,
    callback: saveMemory
  }),

  tool({
    name: 'retrieve_memory',
    description: 'Search and retrieve relevant memories using semantic search.',
    inputSchema: RetrieveMemorySchema,
    callback: retrieveMemory
  })
];